### ISSUE TITLE
Add note on streaming

### ISSUE THREAD

**Issue by rubensworks** [w3c/json-ld-bp#3]
Following https://github.com/w3c/json-ld-bp/issues/4, we can add **guidelines** for achieving streaming JSON-LD, and any **restrictions** it may bring.
Here is a summary of the issues discussed in https://github.com/w3c/json-ld-bp/issues/4.

Guidelines:

1. The structure of a JSON-LD document to enable efficient _streaming parsing_ to RDF.
    1. If there is an `@context` in a node, it should be the first key.
    2. If there is an `@type` in a node, and its value indicates a type-scoped context, it should come right after an `@context` (if there is one), or be the first key.
    3. If there is an `@id` in a node, it should be the first key if there is no `@context` or `@type`, or the second (or third) key if there is an `@context` or `@type`.
2. The order in which RDF triples/quads should appear to enable efficient _streaming serialization_ to JSON-LD.
    1. Quads with equal **graphs** should be grouped (_Achieves grouping of `@graph` blocks_).
    2. Quads with **graph corresponding to the subject** of triples/quads should be grouped (_Achieves grouping of `@graph` and `@id` blocks_).
    3. Triples with equal **subjects** should be grouped (_Achieves grouping of `@id` blocks_).
    4. Triples with equal **predicates** should be grouped (_Achieves grouping of predicate arrays_)

(triple stores may already do these kinds of grouping automatically)

Restrictions:
1. Parsing
    * The spec allows `@context` to appear any in the document. As such, a strict parser may require buffering large portions of the stream. However, since most real-world JSON-LD documents place `@context` as first element, streaming parsers _may_ reject any JSON-LD documents that have out-of-order `@context`'s.
2. Serialization
    * RDF lists are not converted to `@list` arrays, as you can only be certain that `@list` can be used once all triples have been read, which requires keeping the whole stream in-memory.
    * No deduplication of triples, as this would also require keeping the whole stream in-memory.

**Comment by rubensworks** [comment]
Following a comment from @BigBlueHat, scoped `@context` are also allowed, as long as they appear as first key inside their object/scope.

**Comment by gkellogg** [comment]
I think it's only `@type` which would be an issue for ordering and scoped contexts. Individual property's scoped contexts are handled when they are processed, and all context information will already have been processed.

So, the suggested ordering within an object (actually, _node object_ or _graph object_), would be the following:

1. `@context` – if any
2. `@id` or alias – if any
3. `@type` or alias – if any

No other properties (including `@graph`) need be ordered.

**Comment by rubensworks** [comment]
I don't think `@type` should be order-dependent. `@type` essentially expands to `rdf:type`, which makes it processable like any other property/predicate (as long as `@id` comes first).

Unless I'm missing something @gkellogg?

**Comment by gkellogg** [comment]
Scoped contexts can be triggered on `@type` as well as a individual property. If you don’t see `@type` early, you may misinterpret the properties.

**Comment by rubensworks** [comment]
Ah I see, I wasn't aware of that, thanks for clearing that up! ([Link to example for reference](https://w3c.github.io/json-ld-syntax/#example-41-defining-an-context-within-a-term-definition-used-on-type))

**Comment by ajs6f** [comment]
Is this issue distinct from https://github.com/w3c/json-ld-bp/issues/4?

**Comment by rubensworks** [comment]
I would say this is the same as #4. (#4 was moved from another repo)

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2020/2020-02-07-json-ld#section4-1).

- `RESOLVED:  add json-ld-streaming repo; automate all note repos with echidna to publish on each commit`
- `ACTION:  set up json-ld-streaming repo (Ivan Herman)`
<details><summary><i class="differentiate">View the transcript</i></summary>
Streaming best practice<br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-bp/issues/3">https://github.com/w3c/json-ld-bp/issues/3</a><br/>
Ruben Taelman: historically issue 4 was earlier, but in a different tracker and then moved to BP repository<br/>
… I made 3 as a way to summarize the things that are needed to parse efficiently in a streaming manner<br/>
… I think that we can close 4, but I should go through it in more detail to make sure we don’t lose anything<br/>
ajs6f> +1 to closing #4<br/>
Ruben Taelman: that isn’t in #3<br/>
… It would be safe to write some text about those guidelines that I mention in #3, but Gregg appeared to agree with most of it except for the properties<br/>
… the special case where contexts are applied to specific properties, and we would need some more work<br/>
… that we should still look into<br/>
Benjamin Young: Are you up for doing some of the work for the doc?<br/>
Ruben Taelman: Can help with that, but next 2 weeks are very busy<br/>
… only be able to work on it after that, if that’s okay<br/>
Benjamin Young: Some time is better than never! Even just sections and rough pointers for where to flesh out and what you have in mind<br/>
… any content is great and it can be polished later<br/>
… in #4 Adam suggested that this could be a stand alone note, but I don’t think we have time to add a new note<br/>
… It would also need promotion<br/>
Adam Soroka: The impetus behind it was that it’s not part of the specs directly, but we do want to specify precisely<br/>
… in an ideal world it would be nice to have a separate formal note, but not a spec, and the BP doc would just refer to it<br/>
… but whatever mention in the BP doc should be informal rather than pseudo-normative<br/>
… don’t think we have time for a careful spec either, especially in a BP doc<br/>
Benjamin Young: Get what we can into the BP doc<br/>
Adam Soroka: Can potentially publish a note later. Can see what use people make of the notes in the BP doc<br/>
Benjamin Young: Ruben, you mentioned wanting to close #4<br/>
… should we leave that one open?<br/>
… You wanted to go through issue 4 to see what was there<br/>
Ruben Taelman: Assign it to me and I can see if we can close it<br/>
Benjamin Young: Happy to leave opening and closing to you<br/>
… can leave #3 open as the primary topic<br/>
… has a good looking outline<br/>
Ruben Taelman: about the note, I also think it’s valuable to have it at some point. Maybe not in the scope of this WG. Problem I have now is that when I discover JSON-LD docs I want to parse, I have to assume that they are not stream-enabled to be parsed<br/>
… so I parse them the normal way, and can’t use the optimizations for streaming parsing<br/>
… good to add a specific content type or similar to say that it can be parsed in a streaming way<br/>
Benjamin Young: Maybe a profile?<br/>
Rob Sanderson: it seems to ajs6f’s point about formality/informality of the BP…<br/>
… that if we wanted to have event a profile parameter/IRI, that it would raise it to a Note<br/>
Ivan Herman: +1 to Rob<br/>
Rob Sanderson: because we’d want to refer to it from the IANA profile<br/>
… so this does seem like something to consider<br/>
… I agree with rubensworks that if you don’t know you can stream it, you won’t<br/>
… so there needs to be something in the header that tells you it’s possible<br/>
… and perhaps we need to discuss priority of the notes<br/>
Benjamin Young: I think with priority of notes it’s who works on what<br/>
… if they show up then they’ve been prioritized :)<br/>
… if folks want to work on streaming, that’s great. We’re not stealing time from one to the other<br/>
… Have the time to discuss<br/>
… Ivan is this something we can do?<br/>
Ivan Herman: We can add as many notes as we can write<br/>
Ruben Taelman: What are the requirements for such a note?<br/>
… should be more extensive than a section in a BP doc<br/>
Benjamin Young: I think the biggest part is the profile parameters section, but start with whatever you have<br/>
… Is there more plumbing we should set up, Ivan?<br/>
Ivan Herman: In a note we can do what we want.<br/>
Ruben Taelman: Where should it live? Also in the BP repo?<br/>
Benjamin Young: Maybe we should use it as a notes repo<br/>
Ivan Herman: what do you guys want?<br/>
… don’t expect me to make the decision :) To make a new repo is 10 minutes max. You tell me<br/>
Adam Soroka: Was going to ask if we change it to a notes repo, then the BP is a note?<br/>
Benjamin Young: It’s a note already<br/>
… the only one we have content for<br/>
Ivan Herman: the only practical thing, if we want to use things like echidna, it makes it harder if there’s 5 notes in one repo<br/>
… but if we publish only once, it’s not a big deal<br/>
Rob Sanderson: a thought about publishing frequency<br/>
… for a note, having lots and lots of working drafts seems useful to draw attention to it and possibly get input<br/>
… so, I’d prefer we set it up with the best tooling to get drafts into the public<br/>
Ivan Herman: we can set up echidna that every commit to master is auto published<br/>
Benjamin Young: That sounds great<br/>
… a repo for each note<br/>
Ivan Herman: One for BP, one for CBOR. Now there’s a nice bikeshed question … what name for the repo?<br/>
Benjamin Young: json-ld-stream ?<br/>
Benjamin Young: our repos <a href="https://github.com/w3c?utf8=%E2%9C%93&amp;q=json-ld-&type=&language=">https://github.com/w3c?utf8=%E2%9C%93&amp;q=json-ld-&type=&language=</a><br/>
Ivan Herman: +1<br/>
Tim Cole: +1<br/>
Pierre-Antoine Champin: +1<br/>
Proposed resolution: add json-ld-streaming repo; automate all note repos with echidna to publish on each commit (Benjamin Young)<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Adam Soroka: +1<br/>
Ruben Taelman: +1<br/>
Benjamin Young: +1<br/>
Tim Cole: +1<br/>
David I. Lehn: +1<br/>
Ivan Herman: first we have to go tyhrough the pedestrian way, and then we can publish each time<br/>
Benjamin Young: some time in the next couple of weeks we make a drafty working draft<br/>
Resolution #2: add json-ld-streaming repo; automate all note repos with echidna to publish on each commit<br/>
Ivan Herman: that’s something we have a plan to republish CR, and not via echidna but a new CR<br/>
… I have some time constraints about that. I sent a separate note, but I will have a trip to the US at the end of Feb and then a week out 10th of March<br/>
… so the timing might have some consequences<br/>
… that said, the whole procedure has been moved to github, so I believe that chairs can also initiate a FPWD by making an issue<br/>
… and FPWD is more automatic than anything else<br/>
… but still a discussion to have with the webmaster<br/>
Rob Sanderson: if it’s a matter of a week or so, then no need to rush<br/>
Benjamin Young: can always use GH previews for discussion in the mean time<br/>
… ivan do you want an action for the repo?<br/>
Action #1: set up json-ld-streaming repo (Ivan Herman)<br/>
</details>

**Comment by rubensworks** [comment]
Everything mentioned here has now been written here: https://github.com/w3c/json-ld-streaming

So I suggest to close this issue.

**Comment by BigBlueHat** [comment]
Works for me! Thanks, @rubensworks!
